lib/convo_fact.py

```diff
@@ -0,0 +1,309 @@
+import json
+from collections import defaultdict, deque
+from pqdict import PQDict
+from calc_graph import *
+
+class Fact:
+
+    def __init__(self, data):
+        self.id = data['id']
+        self.entities = data['entities']
+        self.question = data['question']
+
+
+class Entity:
+
+    def __init__(self, _id):
+        self.id = _id
+
+
+class CompositeFact:
+
+    def __init__(self, protocol, data):
+        self.id = data['id']
+        self.operator = data['operator']
+        self.facts = [protocol.facts[fact_id] for fact_id in data['facts']]
+
+
+class Protocol:
+
+    def __init__(self, data):
+        self.entities = {}
+        self.facts = {}
+        self.questions = {}
+        self.terminals = []
+
+        for ent in data['entities']:
+            e = Entity(ent)
+            self.entities[ent] = e
+
+        for fact in data['facts']:
+            f = Fact(fact)
+            self.facts[fact['id']] = f
+
+        # topologically sorting composite facts 
+        G = defaultdict(set)
+        composite_facts = {}
+        for composite_fact in data['composite_facts']:
+            composite_facts[composite_fact['id']] = composite_fact
+
+            for fact_id in composite_fact['facts']:
+                G[fact_id].add(composite_fact['id'])
+
+        sequence = []
+
+        def dfs(v):
```

Comment:
need to use loop / stack instead of recursion - explicit recursion depth is limited